---
title: "MapMan2GO - Labbook"
author: "Dr. Asis Hallab & Gabriel Yaxal Ponce Soto"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Step by step workflow used to map MapMan Bins to compound Gene Ontology Term Annotations

Leaf MapMan Bins were provided with their respective identifiers.

### Use UniprotKB Swissprot as reference set.

Curated entries in Swissprot (version as available in May 2018) were annotated with Mercator 4 v.1.0.

### Format the Mercator output to the table that will be used in MapMan2GO.

`cut -f1,3 uniprot_sprot_mercator.tsf | sed "s/'//g" | awk -F'\t' '$2!=""' | tr a-z A-z > SwissProt_Mercator4.txt `

### Prepare UniprotKB Gene Ontology Annotations (GOA) for processing in R

Download `wget ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/goa_uniprot_all.gpa.gz`

Extract important information: UniProtKB gene accessions, evidence codes, and Gene Ontology terms: `awk -F "\t" '/^UniProtKB/ {print $6 "\t" $4 "\t" $2}' goa_uniprot_all.gpa | sort -u > goa_uniprot_all_tbl.txt`.

Reduce the reference GOAs to those genes in relation with MapManBins (see section section "Load MapManBin to Swissprot").

### Prepare protein IDs for taxonomy specific analysis.

Dowload `ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_plants.dat.gz`

Extract protein IDs: 

`awk '{if (/^AC/) print $2}' uniprot_sprot_plants.dat | awk '{gsub (/;/, ""); print $1}' | sort -u >Protein_ID_plants.txt`

### Load MapManBin to Swissprot relations and reduce GOAs to those referenced in the former

Done with R script in this package: `Rscript ./exec/loadSeqSimResults.R path/2/rawMercatorOnSwissprot.tsf path/2/UniProtKB_GOA_preprocessed.txt path/2/MapMan2GO`

The R script saves the resulting data.frames into a binary image (MapManBinsVsSwissprotAndGOA.RData) in this package's data folder.

## Assign the MapMan-Bins compound Gene Ontology Annotations based on the reference annotations found for genes related to the respective Bins

The following R script does the job, saves the resulting data.frames in binary format (MapManBins2GO.RData) in this R package's data folder and generates some plots about the assignment's quality and trustworthyness. A list of protein IDs is provided in order to additionally analyze this subset.

`Rscript ./exec/mapBinsToGOs.R path/2/MapMan2GO path/2/rawMercatorOnSwissprot.tsf [/path/2/SwissProtIDs] [SwissProtIDs_subset_name]`

### Analyze the names of MapMan Bins and GO Terms to identify shared informative words. 

The following R script add the GO Terms which descriptions share words with the MapMan-Bin's description. When identified, the GO terms are added to the GOAs. Saves the resulting data.frames in binary format in the data folder and generates plots about the assignment's quality and information about the GO Terms added. This step can be ignored in order to keep the GOAs only as the intersection.

`Rscript ./exec/addGOstoGOAs.R /path/2/MapMan2GO`

### Calculate and plot the mean, median and maximum values of GO depths.

`Rscript ./exec/plotDepths.R /path/2/MapManBins2GO.RData /path/2/MapMan2GO`

Return plots which are stored in the in the R package.

### Performs a revision on the MapMan Bins - GOAs assignations, classifing GO Terms used and no used as Trusted or Untrusted.

The following R script performs a revision on the MapMan Bins - GOAs assignation, classifying the GO Terms, both, used and no used, as Trusted or Untrusted according to their Evidence Codes (ECO). Returns plots of this analysis.

`Rscript ./exec/mapBinsRevision.R /path/2/MapMan2GO`

## Evaluate Mercator's performance based on MapMan-Bin to GO annotations

Using F1-Scores we compare the GO annotations generated for a reference gene set.
Competetive methods are

- Mercator (MapMan-Bins)
- InterProScan (InterPro2GO)
- Best Blast
- Blasts2GO    (TO DO)

### Extract reference GOAs and InterPro-annotations from Swissprot's GOA

As done above we pre-process the large Swissprot GOA file to extract GO and
InterPro annotations. Then the result is read in R and only those rows are
retained that hold function annotations for our "gold standard" (reference gene
set).

Pre-process goa_uniprot_all.gpa:

`awk -F "\t" '/^UniProtKB/ {if( match($7, /^InterPro/) ){ split($7,a,"|"); for (i in a) { sub(/^InterPro:/, "", a[i]); print $6 "\t" $4 "\t" a[i] "\t" $2 } } else { print  $6 "\t" $4 "\tNA\t" $2 } }' goa_uniprot_all.gpa > goa_uniprot_all_ipr_tbl.txt`

Select a reference genome and extract all gene accessions present in
Uniprot/Swissprot. Here, we use _Arabidopsis thaliana_:

`grep -P '^>' uniprot_sprot.fasta | grep 'OS=Arabidopsis thaliana' | awk -F '|' '{print $2}' > uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt`

Execute the script to extract the reference annotations and store them
in this project's data:

`Rscript ./exec/extractReferenceGoAnnotations.R goa_uniprot_all_ipr_tbl.txt uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt path/2/MapMan2GO`

The performance evaluation of Best Blast was done as described bellow:

0) Generate a protein fasta file containing all Arabidopsis proteins contained in Swissprot

### R-Code

`require(seqinr)`

`sp <- read.fasta( '/path/2/uniprot_sprot.fasta', seqtype='AA', as.string=T, strip.desc=T )`

`options(mc.cores=detectCores())`

`sp.ath.i <- as.logical(unlist(mclapply(sp, function(x) grepl( 'OS=Arabidopsis thaliana', attr(x,"Annot"), fixed=TRUE ))))
sp.ath <- sp[sp.ath.i]`

`write.fasta(sp.ath, names=names(sp.ath), file.out='.i/path/2/store/uniprot_sprot_just_ath.fasta')`

1) Generate a Docker Container with Diamond (Blastp-Alternative)

`docker build -f DiamondAlignerUbuntu -t diamond_aligner_ubuntu:latest .`

2) Create the Diamond database out of the Swissprot curated protein fasta file:

`diamond makedb --in uniprot_sprot.fasta -d uniprot_sprot`

3) Search all the sequences from _Arabidopsis thaliana_ against SwissProt:

`diamond blastp -d uniprot_sprot.dmnd -q uniprot_sprot_just_ath.fasta -o sp_ath_vs_ukb_sprot`

4) Exclude all tautologic matches against Arabidopsis proteins contained in
SwissProt. For this use the Arabidopsis Swissprot Gene Accessions generated
within the MapMan2GO project.

`grep -vf uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt sp_ath_vs_ukb_sprot > sp_ath_vs_ukb_sprot_no_athal.txt`

5) Only retain the best hit for each Arabidopsis query protein

Open an interactive R session, and execute the following code:
### R-Code

`t.v.s <- read.table( './sp_ath_vs_ukb_sprot_no_athal.txt, sep="\t", stringsAsFactors=F )`

`ath.ids <- readLines( './uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt' )`

`t.v.s$id <- sub( "^[^|]+\\|", "", sub( "\\|[^|]+$", "", t.v.s$V2 ) )`

`t.v.s.sort <- t.v.s[ order( t.v.s$V1, t.v.s$V11 ), ]`

`t.v.s.best.hit <- t.v.s.sort[ ! duplicated( t.v.s.sort$V1 ), ]`

`t.v.s.map <- t.v.s.best.hit[ which( t.v.s.best.hit$V3 == 100.0 ), ]`

`write.table( t.v.s.map, './sp_ath_vs_ukb_sprot_best_map.txt', row.names=FALSE, sep="\t", col.names=FALSE, quote=FALSE )`

Finally, execute the script to evaluate the performance of all methodologies

`Rscript ./exec/evalGoPredictionPerformances.R  path/2/mercator_result.txt  path/2/best_blast_preprocessed_results_table.txt  path/2/protein_identifiers_2_exclude  path/2/preprocessed_uniprot_goa_table.txt  path/2/MapMan2GO `

Returns plots of the F1 score for each methodology. 

## Performance evaluation on non SwissProt genes with non electronically made Gene Ontology term annotations

### Obtain gene IDs

We want to obtain gene identifiers that are not contained in SwissProt and do
not have electronically made Gene Ontology term annotations. For this, we use
the Gene Ontology's download section, where sets of GOAs are offered. 

We downloaded the one available for [Oryza
sativa](http://geneontology.org/gene-associations/gene_association.gramene_oryza.gz)
on July 14th 2018. The file is called `gene_association.gramene_oryza` after
unzipping.

Extract those gene UniProtKB gene identifiers meeting the above criteria: (i)
non SwissProt and (ii) non electronically made GOAs (evidence codes not in
(`IEA`, `RCA`, `ISS`).

From the root folder of MapMan2GO:

`Rscript ./exec/extractFromGOANonSprotNon_IEA_ISS_RCA_gene_ids.R gene_association.gramene_oryza uniprot_sprot.fasta .`

The above script will generate a list of UniProtKB short gene identifiers in `inst/nonSwissProtGenesWithNonElectMadeGOAs_IDs.txt`.

In an arbitrary directory of your choice download the UniProtKB fasta amino
acid sequences for these proteins using the UniProt RESTful API:

`cat nonSwissProtGenesWithNonElectMadeGOAs_IDs.txt | awk '{print "wget https://www.uniprot.org/uniprot/" $1 ".fasta"}' | sh`

Then concatonate those files into a single fasta file for downstream analysis:
`cat *.fasta | sed -e '/^\s*$/d' > nonSwissProtGenesWithNonElectMadeGOAs_AA.fasta` 
The result is stored in this package's inst directory.

Subsequently filter out those that have been moved into SwissProt and whose
protein identifier, we supplied the UniProt REST Api is now a synonym. For
this, open an interactive R session and execute the following code from the
root directory of the MapMan2GO R-package:

```
require(seqinr)
res <- read.fasta( './inst/nonSwissProtGenesWithNonElectMadeGOAs_AA.fasta', seqtype='AA', as.string=T, strip.desc=T )
retain.i <- which( grepl( '^tr', names(res), ignore.case=TRUE ) & ! duplicated( names(res) ) )
res.trembl <- res[retain.i]
write.fasta(res.trembl, names=names(res.trembl), file.out='./inst/nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_AA.fasta' )
writeLines( names(res.trembl), './inst/nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_IDs.txt' )
```

### Execute performance analysis on Non Swissprot Non elect annot gold standard

The above gold standard of non SwissProt proteins with non electronically made
GO term annotations is used.

The following three competitors are compared:
* Best Blast (using Diamond) on gold standard searching hits in SwissProt
* Mercator v4 on Gold Standard and using the GO terms mapped to the resulting Bins
* InterProScan and its to GO mappings

The results have been prepared as follows:
* Best Blast: `diamond-aligner blastp -d uniprot_sprot -q nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_AA.fasta -o nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_vs_SwissProt_diamond_out.tsv`
* InterProScan: `/opt/interproscan/interproscan-5.23-62.0/interproscan.sh -i nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_AA.fasta -b Oryza_unique_IPS -goterm` 
* Extract SwissProtID, InterProScanID and GO term from InterProScan output: `cat Oryza_unique_IPS.tsv | grep -P '\bGO:\d{6,6}' | awk -F "\t" '{split($14, a, "|"); for (i in a) { print $1 "\t" $12 "\t" a[i] }}' | sort -u  > nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_interProScan_result.tsv`
* Mercator v4: Uploaded nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_AA.fasta to the Web Interface

Compute performance as follows:

For each of the methods, the F1-Score and Matthews' correlation coefficient were computed. Four evaluations were done: 
(i) Including the whole prediction of the tested method. 
(ii) Including only the GO terms of the prediction belonging to the Biological Process sub-ontology. 
(iii) Including only the GO terms of the prediction belonging to the Cellular Component sub-ontology. 
(iv) Including only the GO terms of the prediction belonging to the Molecular Function sub-ontology.

The following script computes the F1-Scores and Matthews' correlation coeficient of the three methods mentioned above on several levels:
* Considering the universe of all possible GO term annotations without adding ancestral GO terms.
* Considering the universe of all possible GO term annotations including ancestral GO terms.

```
Done with the R script in this package: 
Rscript ./exec/evalGoPredictionPerformances.R inst/nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_mercator_result.tsf inst/nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_vs_SwissProt_diamond_out.tsv inst/nonSwissProtGenesWithNonElectMadeGOAs_trEMBL_interProScan_result.tsv ../goa_uniprot_all_tbl.txt ../performance_evaluation/gene_association.gramene_oryza path/2/MapMan2GO
```
The R script saves the resulting data.frames with the different performances in this package's data folder.

Additional scripts are included to plot performances and save tables of the preformance results.

* generatePlotsPerformanceReport.R
* generateTablesMeanPerformanceReport.R
* generateTextualPerformanceReport.R

The scripts are executed as follow:
```
Rscript. ./exec/generatePlotsPerformanceReport.R path/2/MapMan2GO
```
Generates histograms for precision, recall, false positive rate, F1-Score and Matthews correlation coefficient, as well as scatter plots of false positives, true negatives and number of predictions for Mercator, best Blast and InterProScan performances. 
For each method and each parameter, generates two plots, (i) Including all proteins tested, and (ii) Excluding those proteins with a 100% of secuence similarity with Blast. 
All plots are stored in the inst folder or this package.

```
Rscript. ./exec/generateTablesMeanPerformanceReport.R path/2/MapMan2GO
```
Generate four summary tables of the prediction analysis for the results including: (i) All GO terms, (ii) Biological Process, (iii) Cellular Component and (iv) Molecular Function. The resulting tables are stored in the inst folder of this package.
```
Rscript. ./exec/generateTextualPerformanceReport.R path/2/MapMan2GO
```
Generate a text file containing the performance evaluation of the three methods, including and excluidng the ancestral GO terms, as well as including and excluding those proteins with 100 percent sequence similarity with best Blast. This is done for all the proteins as well as for the three sub-ontologies.

