---
title: "MapMan2GO - Labbook"
author: "Dr. Asis Hallab"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Step by step workflow used to map MapMan Bins to compound Gene Ontology Term Annotations

### Use UniprotKB Swissprot as reference set

Curated entries in Swissprot will be used as reference data. Only 100 percent sequence identity will be accepted. 

#### Run sequence similarity search against Swissprot

Swissprot version is from December 2016. Sequence Similarity Search is carried out with Diamond version (v0.7.9.58).

Generate the database from Swissprot:

`diamond makedb --in uniprot_sprot.fasta -d uniprot_sprot`

Submit job to cluster:

`bsub -q multicore40 -g /dmnd -e ./dmnd.err -o ./dmnd.out -n 10 -R 'span[hosts=1] rusage[mem=128]' 'diamond blastp -q ../MapMan2GO/inst/mapman4.fasta -d ./uniprot_sprot.dmnd -p 10 -a mapman4_vs_sprot -t /scratch/'`

Convert search results to tabular Blast output format: `diamond view -a mapman4_vs_sprot.daa -o mapman4_vs_sprot_blast8.txt`

And retain only matches of 100 percent sequence identity: `awk -F "\t" '{if ($3 == 100.0) { print $1 "\t" $2 }}' mapman4_vs_sprot_blast8.txt | sort -u > mapman4_vs_sprot_blast8_only_100percnt.txt` (Result in this R package's `inst` folder)

### Prepare UniprotKB Gene Ontology Annotations (GOA) for processing in R

Download `wget ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/goa_uniprot_all.gpa.gz`

Extract important information: UniProtKB gene accessions, evidence codes, and Gene Ontology terms: `awk -F "\t" '/^UniProtKB/ {print $6 "\t" $4 "\t" $2}' goa_uniprot_all.gpa | sort -u > goa_uniprot_all_tbl.txt` (Result in this R package's `inst` folder)

Reduceuce the reference GOAs to those genes in relation with MapManBins (see next section).

### Load MapManBin to Swissprot relations and reduce GOAs to those referenced in the former

Done with R script in this package: `Rscript ./exec/loadSeqSimResults.R path/2/preProcessedSeqSimSearchResults.txt path/2/UniProtKB_GOA_preprocessed.txt path/2/MapMan2GO`

The R script saves the resulting data.frames into a binary image in this package's data folder.

## Assign the MapMan-Bins compound Gene Ontology Annotations based on the reference annotations found for genes related to the respective Bins

The following R script does the job, saves the resulting data.frames in binary format in this R package's data folder and generates some plots about the assignment's quality and trustworthyness.

`Rscript ./exec/mapBinsToGOs.R path/2/MapMan2GO`

## Evaluate Mercator's performance based on MapMan-Bin to GO annotations

Using F1-Scores we compare the GO annotations generated for a reference gene set.
Competetive methods are

- Mercator (MapMan-Bins)
- InterProScan (InterPro2GO)
- Best Blast
- Blasts2GO

### Extract reference GOAs and InterPro-annotations from Swissprot's GOA

As done above we pre-process the large Swissprot GOA file to extract GO and
InterPro annotations. Then the result is read in in R and only those rows are
retained that hold function annotations for our "gold standard" (reference gene
set).

Pre-process goa_uniprot_all.gpa:

`awk -F "\t" '/^UniProtKB/ {if( match($7, /^InterPro/) ){ split($7,a,"|"); for (i in a) { sub(/^InterPro:/, "", a[i]); print $6 "\t" $4 "\t" a[i] "\t" $2 } } else { print  $6 "\t" $4 "\tNA\t" $2 } }' goa_uniprot_all.gpa > goa_uniprot_all_ipr_tbl.txt`

Select a reference genome and extract all gene accessions present in
Uniprot/Swissprot. Here, we use _Arabidopsis thaliana_:

`grep -P '^>' uniprot_sprot.fasta | grep 'OS=Arabidopsis thaliana' | awk -F '|' '{print $2}' > uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt`

Finally execute the script to extract the reference annotations and store them
in this project's data:

`Rscript ./exec/extractReferenceGoAnnotations.R goa_uniprot_all_ipr_tbl.txt uniprot_sprot_arabidopsis_thaliana_protein_accessions.txt path/2/MapMan2GO`

